source/common/router/retry_state_impl.cc

```diff
@@ -187,6 +190,13 @@ bool RetryStateImpl::wouldRetry(const Http::HeaderMap* response_headers,
     }
   }
 
+  if (retry_on_ & RetryPolicy::RETRY_ON_GATEWAY_ERROR) {
+    if (response_headers != nullptr &&
```

Comment:
Per your other comment, I think you actually want `||` here for the same reason as the 5xx case. A reset will always result in a 503 or 504 (IIRC for timeout), so should be included.

Reply:
Ah yes, sorry for misunderstanding that. Will update it.